### PR TITLE
Clean up kwargs

### PR DESCRIPTION
**Pull request by mpenkov**
We used to liberally pass arbitrary keyword arguments (e.g. **kwargs). This is a bad practice because it's difficult to keep track of function parameters and ensure they're being passed down to the right function.

This PR starts to clean up our use of arbitrary keyword arguments.

### CODE REVIEW

File: smart_open/smart_open_lib.py [hunk 1]
```diff
@@ -322,8 +322,17 @@ def _shortcut_open(uri, mode, kwargs):
...
     open_kwargs = {}
-    if kwargs.errors is not None:
+    if kwargs.errors is not None and 'b' not in mode:
+        #
+        # Ignore errors parameter if we're reading binary.  This logic is
+        # necessary because:
+        #
+        #   - open complains if you pass non-null errors in binary mode
+        #   - our default mode is rb (incompatible with default Python open)
+        #   - our default errors is non-null (to match default Python open)
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: What does non-null mean? Is null == None?

File: smart_open/smart_open_lib.py [hunk 2]
```diff
@@ -322,8 +322,17 @@ def _shortcut_open(uri, mode, kwargs):
...
 
     open_kwargs = {}
-    if kwargs.errors is not None:
+    if kwargs.errors is not None and 'b' not in mode:
+        #
+        # Ignore errors parameter if we're reading binary.  This logic is
+        # necessary because:
+        #
+        #   - open complains if you pass non-null errors in binary mode
+        #   - our default mode is rb (incompatible with default Python open)
```
Comment (piskvorky) [inline smart_open/smart_open_lib.py]: How about we change this to be compatible?

Or at least make the mode mandatory, so people have to make a conscious decision between binary/text.

File: smart_open/smart_open_lib.py (hunk 1 above)
Comment (mpenkov) [reply smart_open/smart_open_lib.py]: Yes.

File: smart_open/smart_open_lib.py (hunk 2 above)
Comment (mpenkov) [reply smart_open/smart_open_lib.py]: Yes, we will do this in the next major release.

Comment (mpenkov) [reply smart_open/smart_open_lib.py]: BTW, we cannot make the mode mandatory, as that breaks compatibility with the built-in open.

We will switch to reading text mode as default in the next release.

**Comment by menshikh-iv** [comment]
Hi @mpenkov, how is going here? Where you plan to finish this (internal changes + warnings)?

**Comment by mpenkov** [comment]
I'll have a look on the weekend. Thanks for reminding me.

**Comment by mpenkov** [comment]
I figured out a better way to do this.

**Pull request closed without merging**